### ISSUE TITLE
Implicit conversion operator between PointVector

### ISSUE THREAD

**Issue by rolanddenis** [DGtal-team/DGtal#1341]
I'm not sure it is an issue or if we want it to be fixed but, at this time, the `PointVector` conversion operator (from another `PointVector` but with possibly another Euclidean ring type or storage type, see https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/kernel/PointVector.h#L270) is implicit, thus leading to some surprising behaviour:
```C++
using Space = DGtal::SpaceND<2>;
Space::Point a(0, 0); // DGtal::int32_t
Space::RealPoint b(-0.9, -0.9); // double
std::cout << "Is " << a << " lower than " << b << " ? " << ( a.isLower(b) ? "Yes" : "No") << std::endl;
std::cout << "Is " << b << " upper than " << a << " ? " << ( b.isUpper(a) ? "Yes" : "No") << std::endl;
```
outputs:
```
Is [PointVector] {0, 0} lower than [PointVector] {-0.9, -0.9} ? Yes
Is [PointVector] {-0.9, -0.9} upper than [PointVector] {0, 0} ? No
```

It can also happen using integer types only, like (extreme case):
```C++
PointVector<2,DGtal::int8_t>(127, 127).isUpper(PointVector<2,DGtal::int16_t>(128, 128))
```
that is true.

A fix could be: 
- making the conversion operator (that is also the copy constructor) explicit but that generates some compilation errors in DGtal,
- making `isUpper`, `isLower` and other similar methods generic (templated by the euclidean ring type and storage type) so that the comparison would always use the best precision, but some questions can arise : what should `PointVector<int, 2>(1, 1).inf(PointVector<double,2>(0.5,0.5))` return: a `PointVector<int,2>` or a `PointVector<double,2>` ?

An opinion about this ?

**Comment by dcoeurjo** [comment]
So true ;) but you've just opened up Pandora's box ;)

**Comment by JacquesOlivierLachaud** [comment]
Do you have a lot of compilation error in DGtal for option (1). If a few, I would suggest (1) and fixes.
Otherwise, I would go for the implicit "subset" conversion. Since int is a "subset" of double, convert to double. Since float is a subset of double, convert to double, etc.

**Comment by rolanddenis** [comment]
Building tests and examples with default options:
```bash
$ make -j 2>&1 | grep ' error:' | wc -l
166
```
The result of the grep: https://gist.github.com/rolanddenis/eb9a2e3802ce3f1b692499a3f79372e2

I don't think that only allowing "correct" conversion will fixes all the compilation error: considering the example (there is some similar lines among the errors)
```C++
auto pt = PointVector<2,int>(1,2) + PointVector<2,double>(3.5,4.5);
```
it will consider the `operator+` of `PointVector<2,int>` thus failing to find an appropriate conversion from `PointVector<2,double>`. (it maybe works if the operators are declared outside the class, as binary functions but the problem remains for methods like `isUpper`). But that enforce (as the full explicit option) user to take care of conversions !

The second option may look like (choosing best return type):
```C++
template <typename N, typename TL, typename TR, typename CL, typename CR>
typename std::conditional<std::is_same<TL, decltype(TL(0)+TR(0))>, PointVector<N, TL, CL>, PointVector<N, TR, CR>>::type
operator+ ( PointVector<N, TL, CL> const& lhs, PointVector<N, TR, CR> const& rhs )
{
    using ResultPoint = typename std::conditional<std::is_same<TL, decltype(TL(0)+TR(0))>, PointVector<N, TL, CL>, PointVector<N, TR, CR>>::type;
    return ResultPoint(lhs, rhs, std::plus<typename ResultPoint::Component>);
}
```
with conversion constructor still implicit to allow expression like:
```C++
PointVector<2,int> pt = PointVector<2,int>(1,2) + PointVector<2,double>(3.5,4.5) + PointVector<2,double>(0.5,-0.5);
```
but it is not a 5-minutes solution :sweat_smile:

**Comment by dcoeurjo** [comment]
In many situations, the implicit casts (Point->RealPoint) are done on purpose (for instance when drawing stuff or performing some real computations). Some might also come from bugs ;). RealPoint->Point are a bit odd (and probably bugs too).
Long time ago we decided to have the same class for Point and RealPoint for the sake of simplicity (relying on the user/dev smartness to avoid any issue). Splitting the type or controlling the implicit cast is something that could be great (but heavy task).

**Comment by rolanddenis** [comment]
To temper the importance of all those errors : considering errors related to operations between `PointVector`:
```bash
$ grep ' error:' build.log | grep 'operator' | wc -l
94
```
and considering integer to float conversion as valid, even if it may result in a loss of precision, there is "only" 10 errors left (searching for errors where the first operand is not a floating-point type):
```bash
$ grep ' error:' build.log | grep 'operator' | grep -Ev '(float|double).*and' | wc -l
10
```

Doing the same with DtalTools results in 144 errors, including 113 related to operations between `PointVector`, including 2 errors related to float -> integer conversions:
```
/tmp/DGtalTools/converters/mesh2heightfield.cpp:339:83: error: no match for ‘operator+’ (operand types are ‘DGtal::Z3i::Point {aka DGtal::PointVector<3, int, std::array<int, 3> >}’ and ‘DGtal::PointVector<3, double, std::array<double, 3> >’)
/tmp/DGtalTools/converters/vol2heightfield.cpp:222:81: error: no match for ‘operator+’ (operand types are ‘DGtal::Z3i::Point {aka DGtal::PointVector<3, int, std::array<int, 3> >}’ and ‘DGtal::PointVector<3, double, std::array<double, 3> >’)
```

So, allowing such conversions will significantly reduce the compilation errors but using this fix alone will still result is some weird (but valid) behaviours:
```C++
auto pt = PointVector<2,double>(3.5,4.5) + PointVector<2,int>(1,2);
```
would compile but not
```C++
auto pt = PointVector<2,int>(1,2) + PointVector<2,double>(3.5,4.5);
```

**Comment by JacquesOlivierLachaud** [comment]
I am not shocked by the previous behaviour (at least not too much). From an object point of view (and even from a programming language point of view), op::+ is not commutative and result type may differ.

**Comment by rolanddenis** [comment]
But using fundamental types, we are used to have similar results when considering expressions like ```3 + 4.5``` and ```4.5 + 3``` ?

Reconsidering my previous proposed solution, it may need less work than initially guessed:
1. Allow implicit conversion only for the conversion chain `signed char -> short -> int -> long int -> float -> double -> long double` like the standard behaviour for arithmetic operators (see https://en.cppreference.com/w/cpp/language/operator_arithmetic).
2. Explicit conversion otherwise.
3. Allowing other `PointVector` template types for the `PointVector (const Self &apoint1, const UnaryFunctor &f)` constructor thus allowing conversion using a functor (specifying rounding mode for example). Maybe also for the binary version.
4. Adding a construction helper that chooses best `PointVector` type among two given points (following implicit conversion rules) and forwards other parameters to the corresponding constructor, thus hiding the type traits trick proposed before. It may look like (with better name):
    ```C++
    template <typename N, typename TL, typename TR, typename CL, typename CR, typename... Args>
    inline
    typename std::conditional<std::is_same<TL, decltype(TL(0)+TR(0))>, PointVector<N, TL, CL>, PointVector<N, TR, CR>>::type
    constructFromBestPointVectorType( PointVector<N, TL, CL> const&, PointVector<N, TR, CR> const&, Args &&... args)
    {
        using ResultPoint = typename std::conditional<std::is_same<TL, decltype(TL(0)+TR(0))>, PointVector<N, TL, CL>, PointVector<N, TR, CR>>::type;
        return ResultPoint( std::forward<Args>(args)... )
    }
    ```
    This helper could also be used to easily enable implicit conversion only in valid cases (by using a `std::enable_if` based on comparing `Self` type with the result of this helper applied on `Self` and `Other` types.
5. Updating only binary operators of `PointVector`: `+`, `-`, `*`, `/` and comparison operators. It may now look like:
    ```C++
    template <typename N, typename TL, typename TR, typename CL, typename CR>
    auto
    operator+ ( PointVector<N, TL, CL> const& lhs, PointVector<N, TR, CR> const& rhs )
        -> decltype(constructFromBestPointVectorType(lhs, rhs))
    {
        return constructFromBestPointVectorType(lhs, rhs, lhs, rhs, std::plus);
        // Forwards to PointVector (const Self &apoint1, const Self &apoint2, const BinaryFunctor &f) constructor
    }
    ```
6. For the other methods, like `isUpper` and `inf`, no changes so that they work the same if implicit conversion is allowed, otherwise it is the user responsibility to specify the appropriate conversion.

**Comment by JacquesOlivierLachaud** [comment]
Oh !! I didn't know the construction auto f( ... ) -> decltype( g( ... ) )
C++ is so much fun !

**Comment by JacquesOlivierLachaud** [comment]
Otherwise your solution seems pretty neat to me .

**Comment by dcoeurjo** [comment]
Agree.. is it in C++11 already?

**Comment by rolanddenis** [comment]
Yes it is the *trailing return type* syntax introduced in C++11 (see https://en.cppreference.com/w/cpp/language/auto , item (2)). I have also used this syntax in PR #1332 . In C++14, the trailing part (`-> decltype(...)`) can be omitted and the return type will be deduced from the `return` statement.

**Comment by rolanddenis** [comment]
I am also willing to change how comparison operators are implemented: instead of relying on the storage operators, I would explicitly implement them using standard algorithm (`std::equal`, `std::lexicographical_compare`, ...) so that it doesn't depend on the chosen storage. I think it shouldn't change any behaviour since `std::array` comparisons use lexicographical order and I don't have the impression that another storage is currently used ?

An advantage would be to avoid using the conversion constructor before comparing the underlying arrays when conversion is needed.

What do you think ?

**Comment by JacquesOlivierLachaud** [comment]
As far as I know, we always use the defaut comparison operator defined in PointVector. In images, there are subranges with different ordering but they use their own methods.

**Comment by rolanddenis** [comment]
You mean the default storage (thus the lexicographical order) ?

**Comment by JacquesOlivierLachaud** [comment]
Yes. (from std::array)

**Comment by rolanddenis** [comment]
Current draft in #1345. I have add the same conversion rules for the assignment operator, thus triggering a compilation error here: https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/kernel/BasicPointFunctors.h#L459

What should be the wanted conversion ? Round or simply cast to `int` ?

**Comment by dcoeurjo** [comment]
ping @kerautret 
I think rounding should be considered

**Comment by kerautret** [comment]
yes rounding should be ok, sorry

**Comment by rolanddenis** [comment]
Ok, thx !

**Comment by rolanddenis** [comment]
Another case: I have an error here https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/tools/Hull2DHelpers.ih#L445 :
```C++
isInside = (b-a).dot(p)*(b-a).dot(p)>0;
```
since, in this context (used in `tests/geometry/curves/testAlphaThickSegmentComputer.cpp`), `b` and `a` are `PointVector<2, int>` and `p` is a `PointVector<2, double>`. The logical solution is to replace this line by
```C++
isInside = std::pow( p.dot(b-a), 2 ) > 0;
```
by I was wondering if the previous implementation (that is buggy I think) had visible side effects ?

**Comment by JacquesOlivierLachaud** [comment]
Weird line. This line is never negative, so the >0 is just a != 0, no ?

**Comment by rolanddenis** [comment]
Hum wait, it's true :sweat_smile:

But, looking at the documentation, I don't understand what `isInside = (b-a).dot(p)*(b-a).dot(p)>0;` is doing ... should we not simply test if `c[1]` is within `[ a[1], b[1] ]` ?

**Comment by JacquesOlivierLachaud** [comment]
I ping @kerautret because we can blame him ! :-)

**Comment by kerautret** [comment]
oops, yes you are right weird line(s), I miss it on the back port of alpha thick implementation. can at least be replaced by  (b-a).dot(p) >0 I think. sorry!

**Comment by rolanddenis** [comment]
Ok but I don't understand the meaning of this calculation. We are checking if the `p` point, the horizontal projection of `c` on the `(a,b)` line, lies inside the `[a,b]`, right ? If so, we should check `0 <= (b-a).dot(p-a) / (b-a).norm() <= 1` instead, that can be simplified by `a[1] <= c[1] <= b[1]`, no ?

**Comment by kerautret** [comment]
yes, the simplification is ok

**Comment by rolanddenis** [comment]
@kerautret, sorry to bother you again with that but, since the rounding in `Point2DEmbedderIn3D` constructor, the `testBasicPointFunctors` fails:
```
  New Block [Checking projection on a greater space]
    pt [PointVector] {5, 5} => [PointVector] {4, 5, 6} == [PointVector] {4, 5, 6}
    pt [PointVector] {10, 9} => [PointVector] {0, 1, 7} == [PointVector] {0, 1, 7}
    pt [PointVector] {0, 0} => [PointVector] {3, 3, 3} == [PointVector] {3, 3, 3}
    pt [PointVector] {4, 2} => [PointVector] {5, 5, 5} == [PointVector] {5, 5, 5}
    pt [PointVector] {0, 0} => [PointVector] {2, 5, 3} == [PointVector] {3, 5, 3}
    pt [PointVector] {2, 2} => [PointVector] {4, 4, 3} == [PointVector] {5, 4, 3}
    p [PointVector] {5, 2} => [PointVector] {5, 0, 2} == [PointVector] {5, 0, 2}
    p [PointVector] {5, 2} => [PointVector] {2, 5, 0} == [PointVector] {2, 5, 0}
    p [PointVector] {5, 2} => [PointVector] {5, 2, 0} == [PointVector] {5, 2, 0}
    p [PointVector] {5, 2} => [PointVector] {2, -1, -1} == [PointVector] {2, -1, -1}(-1 as default value)
  EndBlock [Checking projection on a greater space] (0.844345 ms)
```
(first error at 5th comparison, corresponding to https://github.com/DGtal-team/DGtal/blob/master/tests/kernel/testBasicPointFunctors.cpp#L143 )

I'm not sure to understand what the correct projections should be in those cases. In fact, using this specific constructor (given a point on the plane and a normal to the plane), I think that the embedding depends on how the plane basis is calculated. So, is the way this basis is actually chosen a kind of canonical way ?

In addition, the fix applied when the `pRefOrigin` point is equal to `anOriginPoint` leads (unless I'm wrong) to a basis that could be not orthogonal to the given normal (eg with an origin at `(1, 0, 0)` and with the normal `(sqrt(2)/2, sqrt(2)/2, 0)`).

(I should maybe open a separate issue ?)

**Comment by kerautret** [comment]
Yes the round should be at the origin of the preview value, the correction to 2->3 should be fine.
For the addition, effectively an issue is perhaps better, I hope to be less busy in the upcoming days but you can assign me ;) thanks in all cases!

**Comment by rolanddenis** [comment]
Ok, I will change the expected results to match the current output and add a comment in order to check it later. Issue opened: #1348 :wink:

**Comment by JacquesOlivierLachaud** [comment]
For me, that kind of expression is error-prone. There are three "natural" readings of this expression, sorted from the most likely to less likely (in my opinion):
1) each component of pt is cast to double and dummy is PointVector<3,double>
2) each component of pt is cast to double, multiplied by 3.5, then rounded again to int and and dummy is PointVector<3,int>
3) each component is multiplied by 3, and dummy is PointVector<3,int>
Since there are several interpretations, I would suggest to forbid that kind of expression. If not forbidden, I would suggest 1) as standard behaviour.

**Comment by dcoeurjo** [comment]
Agree with @JacquesOlivierLachaud.
Actually, only 1. make sense for me. 
If I want 2., I just don't use auto.

**Comment by rolanddenis** [comment]
Ok, I think I will implement the first one but the `dummy` storage type will be the default one (`std::array`), instead of the storage type of one of the parameters (like for `operator+`).

In the same way, I suppose we agree to strictly ban expression like this:
```C++
PointVector<3, int> pt(1, 2, 3);
pt *= 3.5;
```

**Comment by rolanddenis** [comment]
I'm also checking conversion in `dot`, `crossProduct`, etc. About `crossProduct`, while I'm here, I was thinking about enabling this method only for 3D and, eventually, adding a 2D version that only returns the 3rd component of the 3D embedded crossProduct (like in `DGtal::IntegerComputer`). Is it OK ?

**Comment by rolanddenis** [comment]
@kerautret , I also have a conversion issue here https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/curves/AlphaThickSegmentComputer.ih#L756

To fix this, I have supposed that `std::round` was a good choice but can you just confirm me that `InputPoint` will always have an integer component type (so that rounding is always necessary) ?

**Comment by rolanddenis** [comment]
Another question (more for @JacquesOlivierLachaud and @dcoeurjo) : if you look at the first example of the `DistanceBreadthFirstVisitor` documentation page (https://dgtal.org/doc/nightly/classDGtal_1_1DistanceBreadthFirstVisitor.html), there is a `PointVector` conversion issue. The `VertexFunctor` is a composition of a cell embedder (`CanonicSCellEmbedder`) and a distance (`ExactPredicateLpSeparableMetric`) but the first returns a `Space::RealPoint` while the second accepts two `Space::Point` thus leading to unwanted conversions (`double` -> `int`). In addition, if it is actually applied on vertex, the embedder should return points that are not on Z², right ?

Any suggestion ?

**Comment by dcoeurjo** [comment]
hi @rolanddenis, thanks for the issue, it is a deep bug... The Exact* classes should indeed only be used for integer coordinates pont only (otherwise, predicates are not exact at all)... I'll prepare a fix in a PR

**Comment by rolanddenis** [comment]
So far, I found this bug in [`LocalEstimatorFromSurfelFunctorAdapter`](https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/geometry/surfaces/estimation/LocalEstimatorFromSurfelFunctorAdapter.ih#L170) and [`testDistancePropagation.cpp`](https://github.com/DGtal-team/DGtal/blob/master/tests/graph/testDistancePropagation.cpp#L113).

I agree with the use of a lambda, it will be a more easy fix when the `VertexFunctor` type doesn't need to be known in advance.

Note that the update of the `PointVector` class is near the end (except for the `partialCopy` and `partialEqual` that may also be updated) and it remains to fix the triggered conversion issues. I have added some rounding functors to ease the conversion (see https://github.com/DGtal-team/DGtal/pull/1345/commits/1df98efc2181a6e7a74b393cd4394b27ddd1f6e8).

**Comment by rolanddenis** [comment]
Fixed in #1345!
